app/kuma-dp/pkg/dataplane/metrics/server.go

```diff
@@ -120,32 +119,43 @@ func rewriteMetricsURL(path string, port uint32, queryModifier QueryParametersMo
 }
 
 func (s *Hijacker) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
 	out := make(chan []byte, len(s.applicationsToScrape))
 	var wg sync.WaitGroup
+	done := make(chan []byte)
 	wg.Add(len(s.applicationsToScrape))
 	go func() {
 		wg.Wait()
+		close(done)
 		close(out)
 	}()
 
 	for _, app := range s.applicationsToScrape {
 		go func(app ApplicationToScrape) {
 			defer wg.Done()
-			out <- s.getStats(req.URL, app)
+			out <- s.getStats(req.URL, app, ctx)
 		}(app)
 	}
-	for resp := range out {
-		if _, err := writer.Write(resp); err != nil {
-			logger.Error(err, "error while writing the response")
-		}
-		if _, err := writer.Write([]byte("\n")); err != nil {
-			logger.Error(err, "error while writing the response")
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-done:
+		for resp := range out {
+			if _, err := writer.Write(resp); err != nil {
+				logger.Error(err, "error while writing the response")
+			}
+			if _, err := writer.Write([]byte("\n")); err != nil {
+				logger.Error(err, "error while writing the response")
+			}
```

Comment:
Have we considered using: https://pkg.go.dev/golang.org/x/sync@v0.0.0-20220601150217-0de741cfad7f/errgroup ?

Reply:
I am not sure if that is the case here. We don't want to return an error just log it.

Reply:
But the api enables you to do scatter gather using the context like you are doing here no?

Reply:
I guess maybe not because you want partial success to still go forward?

Reply:
Yes, that is the case. If we get only one response we just want to log the failure and return the successful one.